paimon-core/src/main/java/org/apache/paimon/utils/BulkFormatMapping.java

```diff
@@ -139,17 +171,51 @@ public BulkFormatMapping build(
                     indexCastMapping.getIndexMapping(),
                     indexCastMapping.getCastMapping(),
                     partitionMapping,
+                    trimmedKeyPair.getLeft(),
                     formatDiscover
                             .discover(formatIdentifier)
-                            .createReaderFactory(readDataRowType, readFilters),
+                            .createReaderFactory(trimmedKeyPair.getRight(), readFilters),
                     dataSchema,
                     readFilters);
         }
 
-        private List<DataField> readDataFields(TableSchema dataSchema) {
-            List<DataField> dataFields = fieldsExtractor.apply(dataSchema);
+        static Pair<int[], RowType> trimKeyFields(
+                List<DataField> fieldsWithoutPartition, List<DataField> fields) {
+            int[] map = new int[fieldsWithoutPartition.size()];
+            List<DataField> trimmedFields = new ArrayList<>();
+            Map<Integer, DataField> fieldMap = new HashMap<>();
+            Map<Integer, Integer> positionMap = new HashMap<>();
+
+            for (DataField field : fields) {
+                fieldMap.put(field.id(), field);
+            }
+
+            AtomicInteger index = new AtomicInteger();
+            for (int i = 0; i < fieldsWithoutPartition.size(); i++) {
+                DataField field = fieldsWithoutPartition.get(i);
+                boolean keyField = SpecialFields.isKeyField(field.id());
```

Comment:
Why checking this with id? In [Paimon spec](https://paimon.apache.org/docs/master/concepts/spec/datafile/) we state that key column names must start with `_KEY_`. Check their name instead.